### DISCUSSION TITLE
wasm-ld can't recognize .o files. What am I doing wrong?

### DISCUSSION THREAD

**Discussion by 98ahni** [emscripten-core/emscripten#21234] [Q&A] [answered]
I'm trying to compile a project I'm working on but it seems to me that no matter what compiler flags I use I get this error:
`wasm-ld: error: unknown file type: /workspace/bin/intermediate/{any_o_file}.o`

I currently use `emcc "/workspace/Source/{any_cpp_file}.cpp" -D"EMSCRIPTEN=1" -D"__EMSCRIPTEN__=1" -pedantic -x c++ -I"/workspace/Source/" -g -x c++-header -D"DEBUG" -c -O2 -std=c++20 -o /workspace/bin/intermediate/{any_o_file}.o` to create the .o files which works as intended. The error comes when I try to link them with the command `emcc "/workspace/bin/intermediate/file_a.o" "/workspace/bin/intermediate/file_b.o" "/workspace/bin/intermediate/file_c.o" -o /workspace/bin/index.html --bind -O2 --shell-file "/workspace/misc/custom_shell.html" -g3` where wasm-ld complains about all of them being unknown file type. 

What am I missing?

**Comment by sbc100** [comment]
Can you run `file <path/to/object.o>` on one of the input object files to find out what type of file it actually is?

Normally when folks run into this issue it is because they are trying to use ELF object files from the host.

**Comment by sbc100** [comment, marked as answer]
Oh!  I see the problem.  You are using `-x c++-header` which is tell clang to output a C++ pre-compiled header.  

C++ pre-compiled headers are not supposed to be passed to the linker like that and are normally called `.pch` .

**Reply by 98ahni** [reply]
Thank you! 
I would have never thought that to be the problem since it complained about the .o files.

**Comment by sbc100** [comment]
Yup.. the same thing happens on native builds if you try to like a pre-compiled header as an object file.
